model/src/main/proto/math.proto

```diff
@@ -5,17 +5,186 @@ package model;
+  // index). If both this and the end index are zero then no parsing information is included for
+  // this specific expression.
+  uint32 parse_start_index = 1;
+
+  // The index within the input text stream at which point the expression ends, exclusively. If both
+  // this and the start index are zero then no parsing information is included for this specific
+  // expression.
+  uint32 parse_end_index = 2;
+
+  // The type of expression.
+  oneof expression_type {
+    // Indicates that this expression is a real value.
+    Real constant = 3;
+
+    // Indicates that this expression is a variable (which is not valid for numeric-only
+    // expressions).
+    string variable = 4;
+
+    // Indicates that this expression is a binary operation between two sub-expressions.
+    MathBinaryOperation binary_operation = 5;
+
+    // Indicates that this expression is a unary operation that's operating on a sub-expression.
+    MathUnaryOperation unary_operation = 6;
+
+    // Indicates that this expression is a function call with a sub-expression argument.
+    MathFunctionCall function_call = 7;
+
+    // Indicates that this expression represents a nested group, e.g. 1+(2+3).
+    MathExpression group = 8;
+  }
+}
+
+// Represents a binary operation like addition or multiplication.
+//
+// Values of this proto can be analyzed using MathExpressionSubject (within the context of a
+// MathExpression).
+message MathBinaryOperation {
+  // Types of supported binary operations.
+  enum Operator {
+    // Represents an unknown operator (which is never supported).
+    OPERATOR_UNSPECIFIED = 0;
+
+    // Represents adding two values, e.g.: 1+x.
+    ADD = 1;
+
+    // Represents subtracting two values, e.g.: x-2.
+    SUBTRACT = 2;
+
+    // Represents multiplying two values, e.g.: x*y.
+    MULTIPLY = 3;
+
+    // Represents dividing two values, e.g.: 1/x.
+    DIVIDE = 4;
+
+    // Represents taking the exponentiation of one value by another, e.g.: x^2.
+    EXPONENTIATE = 5;
+  }
+
+  // The type of binary operation.
+  Operator operator = 1;
+
+  // The left-hand side of the operation, e.g. the '1' in 1+2.
+  MathExpression left_operand = 2;
+
+  // The right-hand side of the operation, e.g. the '2' in 1+2.
+  MathExpression right_operand = 3;
+
+  // Indicates whether this operation is implicit. This is currently only supported for
+  // multiplication, and helps represent expressions like '2x' (which should be treated as 2*x).
+  bool is_implicit = 4;
+}
+
+// Represents a unary operation like negation.
+//
+// Values of this proto can be analyzed using MathExpressionSubject (within the context of a
+// MathExpression).
+message MathUnaryOperation {
+  // Types of supported unary operations.
+  enum Operator {
+    // Represents an unknown operator (which is never supported).
+    OPERATOR_UNSPECIFIED = 0;
+
+    // Represents negating a value, e.g.: -y.
+    NEGATE = 1;
+
+    // Represents indicating a value as positive, e.g.: +y.
+    POSITIVE = 2;
```

Comment:
Do we actually need this? Can we not say that anything thats not negative will be a positive?

Reply:
The grammar defines positives as valid expressions for three reasons:
1. This is conventional with other calculators (such as the Ubuntu calculator, Google's, and Wolfram Alpha)
2. It provides parity with unary plus (consider the parity between addition & subtraction)
3. It makes it more straightforward to cleanly detect this as an error case (since we don't actually want learners to do this since it's likely an incomplete addition operation). Rather than special casing a specific invalid token case we can do a post-parsing analysis for this type.

I suggest looking at the tests in #4052 if you want to see examples where unary plus is used.